File: csharp/src/Microsoft.Azure.EventHubs/Primitives/ConnectionStringBuilder.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

namespace Microsoft.Azure.EventHubs
{
    using System;
    using System.Text;

    /// <summary>
    /// ConnectionStringBuilder can be used to construct a connection string which can establish communication with ServiceBus entities.
    /// It can also be used to perform basic validation on an existing connection string.
    /// <para/>
    /// A connection string is basically a string consisted of key-value pair separated by ";". 
    /// Basic format is "&lt;key&gt;=&lt;value&gt;[;&lt;key&gt;=&lt;value&gt;]" where supported key name are as follow:
    /// <para/> Endpoint - the URL that contains the servicebus namespace
    /// <para/> EntityPath - the path to the service bus entity (queue/topic/eventhub/subscription/consumergroup/partition)
    /// <para/> SharedAccessKeyName - the key name to the corresponding shared access policy rule for the namespace, or entity.
    /// <para/> SharedAccessKey - the key for the corresponding shared access policy rule of the namespace or entity.
    /// </summary>
    /// <example>
    /// Sample code:
    /// <code>
    /// ConnectionStringBuilder connectionStringBuilder = new ConnectionStringBuilder(
    ///     "ServiceBusNamespaceName", 
    ///     "ServiceBusEntityName", // eventHub, queue, or topic name 
    ///     "SharedAccessSignatureKeyName", 
    ///     "SharedAccessSignatureKey");
    ///  string connectionString = connectionStringBuilder.toString();
    /// </code>
    /// </example>
    public class ConnectionStringBuilder
    {
        static readonly TimeSpan DefaultOperationTimeout = TimeSpan.FromMinutes(1);
        static readonly string EndpointFormat = "amqps://{0}.servicebus.windows.net";
        static readonly string EndpointConfigName = "Endpoint";
        static readonly string SharedAccessKeyNameConfigName = "SharedAccessKeyName";
        static readonly string SharedAccessKeyConfigName = "SharedAccessKey";
        static readonly string EntityPathConfigName = "EntityPath";
        const char KeyValueSeparator = '=';
        const char KeyValuePairDelimiter = ';';

        /// <summary>
        /// Build a connection string consumable by <see cref="EventHubClient.Create(string)"/>
        /// </summary>
        /// <param name="namespaceName">Namespace name (the dns suffix, ex: .servicebus.windows.net, is not required)</param>
        /// <param name="entityPath">Entity path. For eventHubs case specify eventHub name.</param>
        /// <param name="sharedAccessKeyName">Shared Access Key name</param>
        /// <param name="sharedAccessKey">Shared Access Key</param>
        public ConnectionStringBuilder(string namespaceName, string entityPath, string sharedAccessKeyName, string sharedAccessKey)
            : this(namespaceName, entityPath, sharedAccessKeyName, sharedAccessKey, DefaultOperationTimeout, RetryPolicy.Default)
        {
        }

        ConnectionStringBuilder(
            string namespaceName,
            string entityPath,
            string sharedAccessKeyName,
            string sharedAccessKey,
            TimeSpan operationTimeout,
            RetryPolicy retryPolicy)
        {
            if (string.IsNullOrWhiteSpace(namespaceName) || string.IsNullOrWhiteSpace(entityPath))
            {
                throw ExceptionUtility.ArgumentNullOrWhiteSpace(string.IsNullOrWhiteSpace(namespaceName) ? nameof(namespaceName) : nameof(entityPath));
            }
            else if (string.IsNullOrWhiteSpace(sharedAccessKeyName) || string.IsNullOrWhiteSpace(sharedAccessKey))
            {
                throw ExceptionUtility.ArgumentNullOrWhiteSpace(string.IsNullOrWhiteSpace(sharedAccessKeyName) ? nameof(sharedAccessKeyName) : nameof(sharedAccessKey));
            }
            else if (retryPolicy == null)
            {
                throw ExceptionUtility.ArgumentNull(nameof(retryPolicy));
            }

            this.Endpoint = new Uri(EndpointFormat.FormatInvariant(namespaceName));
            this.EntityPath = entityPath;
            this.SasKey = sharedAccessKey;
            this.SasKeyName = sharedAccessKeyName;
            this.OperationTimeout = operationTimeout;
            this.RetryPolicy = retryPolicy;
        }

        /// <summary>
        /// ConnectionString format:
        /// Endpoint=sb://namespace_DNS_Name;EntityPath=EVENT_HUB_NAME;SharedAccessKeyName=SHARED_ACCESS_KEY_NAME;SharedAccessKey=SHARED_ACCESS_KEY
        /// </summary>
        /// <param name="connectionString">ServiceBus ConnectionString</param>
        public ConnectionStringBuilder(string connectionString)
        {
            if (string.IsNullOrWhiteSpace(connectionString))
            {
                throw ExceptionUtility.ArgumentNullOrWhiteSpace(nameof(connectionString));
            }

            this.ParseConnectionString(connectionString);
        }

        public Uri Endpoint { get; set; }

        /// <summary>
        /// Get the shared access policy key value from the connection string
        /// </summary>
        /// <value>Shared Access Signature key</value>
        public string SasKey { get; set; }

        /// <summary>
        /// Get the shared access policy owner name from the connection string
        /// </summary>
        public string SasKeyName { get; set; }

        /// <summary>
        /// Get the entity path value from the connection string
        /// </summary>
        public string EntityPath { get; set; }

        /// <summary>
        /// OperationTimeout is applied in erroneous situations to notify the caller about the relevant <see cref="ServiceBusException"/>
        /// </summary>
        public TimeSpan OperationTimeout { get; set; }

        /// <summary>
        /// Get the retry policy instance that was created as part of this builder's creation.
        /// </summary>
        public RetryPolicy RetryPolicy { get; set; }

        /// <summary>
        /// Returns an interoperable connection string that can be used to connect to ServiceBus Namespace
        /// </summary>
        /// <returns>the connection string</returns>
        public override string ToString()
        {
            var connectionStringBuilder = new StringBuilder();
            if (this.Endpoint != null)
            {
                connectionStringBuilder.Append($"{EndpointConfigName}{KeyValueSeparator}{this.Endpoint}{KeyValuePairDelimiter}");
            }

            if (!string.IsNullOrWhiteSpace(this.EntityPath))
            {
                connectionStringBuilder.Append($"{EntityPathConfigName}{KeyValueSeparator}{this.EntityPath}{KeyValuePairDelimiter}");
            }

            if (!string.IsNullOrWhiteSpace(this.SasKeyName))
            {
                connectionStringBuilder.Append($"{SharedAccessKeyNameConfigName}{KeyValueSeparator}{this.SasKeyName}{KeyValuePairDelimiter}");
            }

            if (!string.IsNullOrWhiteSpace(this.SasKey))
            {
                connectionStringBuilder.Append($"{SharedAccessKeyConfigName}{KeyValueSeparator}{this.SasKey}");
            }

            return connectionStringBuilder.ToString();
        }

        internal EventHubClient CreateEventHubClient()
        {
            // In the future to support other protocols add that logic here.
            return new AmqpEventHubClient(this);
        }

        void ParseConnectionString(string connectionString)
        {
            // First split based on ';'
            string[] keyValuePairs = connectionString.Split(new[] { KeyValuePairDelimiter }, StringSplitOptions.RemoveEmptyEntries);
            foreach (var keyValuePair in keyValuePairs)
            {
                // Now split based on the _first_ '='
                string[] keyAndValue = keyValuePair.Split(new[] { KeyValueSeparator }, 2);
                string key = keyAndValue[0];
                if (keyAndValue.Length != 2)
                {
                    throw ExceptionUtility.Argument(nameof(connectionString), $"Value for the connection string parameter name '{key}' was not found.");
                }

                string value = keyAndValue[1];
                if (key.Equals(EndpointConfigName, StringComparison.OrdinalIgnoreCase))
                {
                    this.Endpoint = new Uri(value);
                }
                else if (key.Equals(EntityPathConfigName, StringComparison.OrdinalIgnoreCase))
                {
                    this.EntityPath = value;
                }
                else if (key.Equals(SharedAccessKeyNameConfigName, StringComparison.OrdinalIgnoreCase))
                {
                    this.SasKeyName = value;
                }
                else if (key.Equals(SharedAccessKeyConfigName, StringComparison.OrdinalIgnoreCase))
                {
                    this.SasKey = value;
                }
                else
                {
                    throw ExceptionUtility.Argument(nameof(connectionString), $"Illegal connection string parameter name '{key}'");
                }
            }
        }
    }
}

File: java/azure-eventhubs/src/main/java/com/microsoft/azure/servicebus/ConnectionStringBuilder.java
/*
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project root for full license information.
 */
package com.microsoft.azure.servicebus;

import java.net.*;
import java.time.*;
import java.time.format.DateTimeParseException;
import java.util.*;
import java.util.regex.*;
import com.microsoft.azure.eventhubs.*;

/**
 * {@link ConnectionStringBuilder} can be used to construct a connection string which can establish communication with ServiceBus entities.
 * It can also be used to perform basic validation on an existing connection string.
 * <p> Sample Code:
 * <pre>{@code
 * 	ConnectionStringBuilder connectionStringBuilder = new ConnectionStringBuilder(
 *     "ServiceBusNamespaceName", 
 *     "ServiceBusEntityName", // eventHubName or QueueName or TopicName 
 *     "SharedAccessSignatureKeyName", 
 *     "SharedAccessSignatureKey");
 *  
 * String connectionString = connectionStringBuilder.toString();
 * }</pre>
 * <p>
 * A connection string is basically a string consisted of key-value pair separated by ";". 
 * Basic format is {{@literal <}key{@literal >}={@literal <}value{@literal >}[;{@literal <}key{@literal >}={@literal <}value{@literal >}]} where supported key name are as follow:
 * <ul>
 * <li> Endpoint - the URL that contains the servicebus namespace
 * <li> EntityPath - the path to the service bus entity (queue/topic/eventhub/subscription/consumergroup/partition)
 * <li> SharedAccessKeyName - the key name to the corresponding shared access policy rule for the namespace, or entity.
 * <li> SharedAccessKey - the key for the corresponding shared access policy rule of the namespace or entity.
 * </ul>
 */
public class ConnectionStringBuilder
{
	final static String endpointFormat = "amqps://%s.servicebus.windows.net";
	
	final static String EndpointConfigName = "Endpoint";
	final static String SharedAccessKeyNameConfigName = "SharedAccessKeyName";
	final static String SharedAccessKeyConfigName = "SharedAccessKey";
	final static String EntityPathConfigName = "EntityPath";
	final static String OperationTimeoutConfigName = "OperationTimeout";
	final static String KeyValueSeparator = "=";
	final static String KeyValuePairDelimiter = ";";

	private static final String AllKeyEnumerateRegex = "(" + EndpointConfigName + "|" + SharedAccessKeyNameConfigName
			+ "|" + SharedAccessKeyConfigName + "|" + EntityPathConfigName + "|" + OperationTimeoutConfigName + ")";

	private static final String KeysWithDelimitersRegex = KeyValuePairDelimiter + AllKeyEnumerateRegex
			+ KeyValueSeparator;

	private String connectionString;
	private URI endpoint;
	private String sharedAccessKeyName;
	private String sharedAccessKey;
	private String entityPath;
	private Duration operationTimeout;
	private RetryPolicy retryPolicy;

	private ConnectionStringBuilder(final String namespaceName, final String entityPath, final String sharedAccessKeyName,
			final String sharedAccessKey, final Duration operationTimeout, final RetryPolicy retryPolicy)
	{
		try
		{
			this.endpoint = new URI(String.format(Locale.US, endpointFormat, namespaceName));
		} 
		catch(URISyntaxException exception)
		{
			throw new IllegalConnectionStringFormatException(
					String.format(Locale.US, "Invalid namespace name: %s", namespaceName),
					exception);
		}
		
		this.sharedAccessKey = sharedAccessKey;
		this.sharedAccessKeyName = sharedAccessKeyName;
		this.operationTimeout = operationTimeout;
		this.retryPolicy = retryPolicy;
		this.entityPath = entityPath;
	}

	/**
	 * Build a connection string consumable by {@link EventHubClient#createFromConnectionString(String)}
	 * @param namespaceName Namespace name (dns suffix - ex: .servicebus.windows.net is not required) 
	 * @param entityPath Entity path. For eventHubs case specify - eventHub name.
	 * @param sharedAccessKeyName Shared Access Key name
	 * @param sharedAccessKey Shared Access Key
	 */
	public ConnectionStringBuilder(final String namespaceName, final String entityPath, final String sharedAccessKeyName,
			final String sharedAccessKey)
	{
		this(namespaceName, entityPath, sharedAccessKeyName, sharedAccessKey, MessagingFactory.DefaultOperationTimeout, RetryPolicy.getDefault());
	}

	/**
	 * ConnectionString format:
	 * 		Endpoint=sb://namespace_DNS_Name;EntityPath=EVENT_HUB_NAME;SharedAccessKeyName=SHARED_ACCESS_KEY_NAME;SharedAccessKey=SHARED_ACCESS_KEY
	 * @param connectionString ServiceBus ConnectionString
	 * @throws IllegalConnectionStringFormatException when the format of the ConnectionString is not valid
	 */
	public ConnectionStringBuilder(String connectionString)
	{
		this.parseConnectionString(connectionString);
		this.connectionString = connectionString;
	}
	
	/**
	 * Get the endpoint which can be used to connect to the ServiceBus Namespace
	 * @return Endpoint
	 */
	public URI getEndpoint()
	{
		return this.endpoint;
	}

	/**
	 * Get the shared access policy key value from the connection string
	 * @return Shared Access Signature key
	 */
	public String getSasKey()
	{
		return this.sharedAccessKey;
	}

	/**
	 * Get the shared access policy owner name from the connection string
	 * @return Shared Access Signature key name.
	 */
	public String getSasKeyName()
	{
		return this.sharedAccessKeyName;
	}
	
	/**
	 * Get the entity path value from the connection string
	 * @return Entity Path
	 */
	public String getEntityPath()
	{
		return this.entityPath;
	}
	
	/**
	 * OperationTimeout is applied in erroneous situations to notify the caller about the relevant {@link ServiceBusException}
	 * @return operationTimeout
	 */
	public Duration getOperationTimeout()
	{
		return (this.operationTimeout == null ? MessagingFactory.DefaultOperationTimeout : this.operationTimeout);
	}
	
	/**
	 * Set the OperationTimeout value in the Connection String. This value will be used by all operations which uses this {@link ConnectionStringBuilder}, unless explicitly over-ridden.
	 * @param operationTimeout Operation Timeout
	 */
	public void setOperationTimeout(final Duration operationTimeout)
	{
		this.operationTimeout = operationTimeout;
	}
	
	/**
	 * Get the retry policy instance that was created as part of this builder's creation.
	 * @return RetryPolicy applied for any operation performed using this ConnectionString
	 */
	public RetryPolicy getRetryPolicy()
	{
		return (this.retryPolicy == null ? RetryPolicy.getDefault() : this.retryPolicy);
	}

	/**
	 * Returns an inter-operable connection string that can be used to connect to ServiceBus Namespace
	 * @return connection string
	 */
	@Override
	public String toString()
	{
		if (StringUtil.isNullOrWhiteSpace(this.connectionString))
		{
			StringBuilder connectionStringBuilder = new StringBuilder();
			if (this.endpoint != null)
			{
				connectionStringBuilder.append(String.format(Locale.US, "%s%s%s%s", EndpointConfigName, KeyValueSeparator,
					this.endpoint.toString(), KeyValuePairDelimiter));
			}
			
			if (!StringUtil.isNullOrWhiteSpace(this.entityPath))
			{
				connectionStringBuilder.append(String.format(Locale.US, "%s%s%s%s", EntityPathConfigName,
					KeyValueSeparator, this.entityPath, KeyValuePairDelimiter));
			}
			
			if (!StringUtil.isNullOrWhiteSpace(this.sharedAccessKeyName))
			{
				connectionStringBuilder.append(String.format(Locale.US, "%s%s%s%s", SharedAccessKeyNameConfigName,
					KeyValueSeparator, this.sharedAccessKeyName, KeyValuePairDelimiter));
			}
			
			if (!StringUtil.isNullOrWhiteSpace(this.sharedAccessKey))
			{
				connectionStringBuilder.append(String.format(Locale.US, "%s%s%s", SharedAccessKeyConfigName,
					KeyValueSeparator, this.sharedAccessKey));
			}
			
			if (this.operationTimeout != null)
			{
				connectionStringBuilder.append(String.format(Locale.US, "%s%s%s%s", KeyValuePairDelimiter, OperationTimeoutConfigName,
						KeyValueSeparator, this.operationTimeout.toString()));
			}
			
			this.connectionString = connectionStringBuilder.toString();
		}

		return this.connectionString;
	}

	private void parseConnectionString(String connectionString)
	{
		// TODO: Trace and throw
		if (StringUtil.isNullOrWhiteSpace(connectionString))
		{
			throw new IllegalConnectionStringFormatException(String.format("connectionString cannot be empty"));
		}

		String connection = KeyValuePairDelimiter + connectionString;

		Pattern keyValuePattern = Pattern.compile(KeysWithDelimitersRegex, Pattern.CASE_INSENSITIVE);
		String[] values = keyValuePattern.split(connection);
		Matcher keys = keyValuePattern.matcher(connection);

		if (values == null || values.length <= 1 || keys.groupCount() == 0)
		{
			throw new IllegalConnectionStringFormatException("Connection String cannot be parsed.");
		}

		if (!StringUtil.isNullOrWhiteSpace((values[0])))
		{
			throw new IllegalConnectionStringFormatException(
					String.format(Locale.US, "Cannot parse part of ConnectionString: %s", values[0]));
		}

		int valueIndex = 0;
		while (keys.find())
		{
			valueIndex++;

			String key = keys.group();
			key = key.substring(1, key.length() - 1);
			
			if (values.length < valueIndex + 1)
			{
				throw new IllegalConnectionStringFormatException(
						String.format(Locale.US, "Value for the connection string parameter name: %s, not found", key));
			}
			
			if (key.equalsIgnoreCase(EndpointConfigName))
			{
				try
				{
					this.endpoint = new URI(values[valueIndex]); 
				}
				catch(URISyntaxException exception)
				{
					throw new IllegalConnectionStringFormatException(
							String.format(Locale.US, "%s should be in format scheme://fullyQualifiedServiceBusNamespaceEndpointName", EndpointConfigName),
							exception);
				}
			}
			else if(key.equalsIgnoreCase(SharedAccessKeyNameConfigName))
			{
				this.sharedAccessKeyName = values[valueIndex];
			}
			else if(key.equalsIgnoreCase(SharedAccessKeyConfigName))
			{
				this.sharedAccessKey = values[valueIndex];
			}
			else if (key.equalsIgnoreCase(EntityPathConfigName))
			{
				this.entityPath = values[valueIndex];
			}
			else if (key.equalsIgnoreCase(OperationTimeoutConfigName))
			{
				try
				{
					this.operationTimeout = Duration.parse(values[valueIndex]);
				}
				catch(DateTimeParseException exception)
				{
					throw new IllegalConnectionStringFormatException("Invalid value specified for property 'Duration' in the ConnectionString.", exception);
				}
			}
			else
			{
				throw new IllegalConnectionStringFormatException(
						String.format(Locale.US, "Illegal connection string parameter name: %s", key));
			}
		}
	}
}
